src/utils/grpc/grpc-client.ts

```diff
@@ -100,50 +99,70 @@ export type GRPCClusterMethods = {
   ) => Promise<RequestCancelWorkflowExecutionResponse>;
 };
 
-const getClusterServices = async () => {
-  const CLUSTERS_CONFIGS = await getConfigValue('CLUSTERS');
-  return CLUSTERS_CONFIGS.reduce((result, c) => {
-    const requestConfig: GRPCRequestConfig = {
-      serviceName: c.grpc.serviceName,
-      metadata: c.grpc.metadata,
-    };
+// cache services instances
+const clisterServicesMapGlobalRef = new GlobalRef<ClustersServices>(
+  'cluster-services-map',
+  {}
+);
+const clusterServicesMap: ClustersServices = clisterServicesMapGlobalRef.value;
+
+const getClusterServices = async (c: ClusterConfig) => {
+  if (clusterServicesMap[c.clusterName]) {
+    return clusterServicesMap[c.clusterName];
+  }
+
+  const requestConfig: GRPCRequestConfig = {
+    serviceName: c.grpc.serviceName,
+    metadata: c.grpc.metadata,
+  };
 
-    const adminService = new GRPCService({
-      peer: c.grpc.peer,
-      requestConfig,
-      ...grpcServiceConfigurations.adminServiceConfig,
-    });
-    const domainService = new GRPCService({
-      peer: c.grpc.peer,
-      requestConfig,
-      ...grpcServiceConfigurations.domainServiceConfig,
-    });
-    const visibilityService = new GRPCService({
-      peer: c.grpc.peer,
-      requestConfig,
-      ...grpcServiceConfigurations.visibilityServiceConfig,
-    });
-    const workflowService = new GRPCService({
-      peer: c.grpc.peer,
-      requestConfig,
-      ...grpcServiceConfigurations.workflowServiceConfig,
-    });
+  const adminService = new GRPCService({
+    peer: c.grpc.peer,
+    requestConfig,
+    ...grpcServiceConfigurations.adminServiceConfig,
+  });
+  const domainService = new GRPCService({
+    peer: c.grpc.peer,
+    requestConfig,
+    ...grpcServiceConfigurations.domainServiceConfig,
+  });
+  const visibilityService = new GRPCService({
+    peer: c.grpc.peer,
+    requestConfig,
+    ...grpcServiceConfigurations.visibilityServiceConfig,
+  });
+  const workflowService = new GRPCService({
+    peer: c.grpc.peer,
+    requestConfig,
+    ...grpcServiceConfigurations.workflowServiceConfig,
+  });
 
-    result[c.clusterName] = {
-      adminService,
-      domainService,
-      visibilityService,
-      workflowService,
-    };
-    return result;
-  }, {} as ClusterServices);
+  const services: ClusterService = {
+    adminService,
+    domainService,
+    visibilityService,
+    workflowService,
+  };
+  // add service to cache (clusterServicesMap)
+  clusterServicesMap[c.clusterName] = services;
+  return services;
+};
+
+const getClustersServices = async () => {
```

Comment:
nit: Would suggest renaming this to getAllClustersServices or something similar just for clarity.